src/hooks/useEditorView.ts

```diff
@@ -59,31 +36,66 @@ function withFlushedDispatch(
  * All state and props updates are executed in a layout effect.
  * To ensure that the EditorState and EditorView are never out of
  * sync, it's important that the EditorView produced by this hook
- * is only accessed through the `useEditorViewEvent` and
- * `useEditorViewLayoutEffect` hooks.
+ * is only accessed through the provided hooks.
  */
 export function useEditorView<T extends HTMLElement = HTMLElement>(
   mount: T | null,
-  props: EditorProps
-): EditorView | null {
-  const [view, setView] = useState<EditorView | null>(null);
-  const forceUpdate = useForceUpdate();
-
-  const editorProps = withFlushedDispatch(props, forceUpdate);
+  props: Props
+): EditorContextValue {
+  if (process.env.NODE_ENV !== "production") {
+    if (
+      props.defaultState !== undefined &&
+      props.state !== undefined &&
+      !didWarnValueDefaultValue
+    ) {
+      console.error(
+        "A component contains a ProseMirror editor with both value and defaultValue props. " +
+          "ProseMirror editors must be either controlled or uncontrolled " +
+          "(specify either the state prop, or the defaultState prop, but not both). " +
+          "Decide between using a controlled or uncontrolled ProseMirror editor " +
+          "and remove one of these props. More info: " +
+          "https://reactjs.org/link/controlled-components"
+      );
+      didWarnValueDefaultValue = true;
+    }
+  }
 
-  const stateProp = "state" in editorProps ? editorProps.state : undefined;
+  const defaultState = props.defaultState ?? EMPTY_STATE;
+  const [_state, setState] = useState<EditorState>(defaultState);
+  const state = props.state ?? _state;
 
-  const state =
-    "defaultState" in editorProps
-      ? editorProps.defaultState
-      : editorProps.state;
+  const {
+    componentEventListenersPlugin,
+    registerEventListener,
+    unregisterEventListener,
+  } = useComponentEventListeners();
 
-  const nonStateProps = Object.fromEntries(
-    Object.entries(editorProps).filter(
-      ([propName]) => propName !== "state" && propName !== "defaultState"
-    )
+  const plugins = useMemo(
+    () => [...(props.plugins ?? []), componentEventListenersPlugin],
+    [props.plugins, componentEventListenersPlugin]
   );
 
+  function dispatchTransaction(this: EditorView, tr: Transaction) {
+    flushSync(() => {
+      if (!props.state) {
+        setState((s) => s.apply(tr));
+      }
+
+      if (props.dispatchTransaction) {
+        props.dispatchTransaction.call(this, tr);
+      }
+    });
+  }
```

Comment:
This does still have the potentially weird effect of basically overriding a consumers' `updateState()` in `dispatchTransaction()` (since we explicitly set the state on the view later on, in the layout effect), but I think that's probably the most correct outcome (and I would still like to think about warning about it in the future I think).

Reply:
I don't think we have to do anything about this, just making note of it!

Reply:
Yup. I think we might be able to defer this a bit. I'd like to see how things look once we have the pure React view implementation. Maybe then we can actually store the view's state in React or something.

Reply:
Yeah this definitely feels less fraught in the react-editor-view branch, since we don't have to use layout effects to update the state in the view!